predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -875,3 +887,92 @@ def _get_successors(
         score_function.evaluate(kept_predicates)  # log useful numbers
 
         return set(kept_predicates)
+
+    def _select_predicates_by_clustering(
+            self, candidates: Dict[Predicate, float],
+            initial_predicates: Set[Predicate], dataset: Dataset,
+            atom_dataset: List[GroundAtomTrajectory]) -> Set[Predicate]:
+        """Cluster segments from the atom_dataset into clusters corresponding
+        to operators and use this to select predicates."""
+
+        if CFG.grammar_search_pred_clusterer == "oracle":
+            assert CFG.offline_data_method == "demo+gt_operators"
+            assert dataset.annotations is not None and len(
+                dataset.annotations) == len(dataset.trajectories)
+            assert CFG.segmenter == "option_changes"
+            segmented_trajs = [
+                segment_trajectory(traj) for traj in atom_dataset
+            ]
+            assert len(segmented_trajs) == len(dataset.annotations)
+            # First, get the set of all ground truth operator names.
+            all_gt_op_names = set(ground_nsrt.parent.name
+                                  for anno_list in dataset.annotations
+                                  for ground_nsrt in anno_list)
+            # Next, make a dictionary mapping operator name to segments
+            # where that operator was used.
+            gt_op_to_segments: Dict[str, List[Segment]] = {
+                op_name: []
+                for op_name in all_gt_op_names
+            }
+            for op_list, seg_list in zip(dataset.annotations, segmented_trajs):
+                assert len(seg_list) == len(op_list)
+                for ground_nsrt, segment in zip(op_list, seg_list):
+                    gt_op_to_segments[ground_nsrt.parent.name].append(segment)
+            consistent_add_effs_preds: Set[Predicate] = set()
+            # First, select predicates that change as add effects consistently
```

Comment:
this and the below feels redundant to me. can we just check for consistency once? for example, can we just use the below code, and perhaps update it slightly to remove any predicates that never change?